src/presentation.ts

```diff
@@ -14,178 +22,223 @@ import {
+} from './deserialize-spec.ts';
 
+// external API
+export { PresentationRequest, Presentation };
+
+// internal
 export {
-  PresentationRequest,
-  Presentation,
   type ZkAppInputContext,
   type HttpsInputContext,
+  type WalletDerivedContext,
+  ZkAppRequest,
+  HttpsRequest,
+  hashClaims,
 };
 
-type BaseInputContext = {
-  presentationCircuitVKHash: Field;
-  serverNonce: Field;
-};
-
-type ZkAppInputContext = BaseInputContext & {
-  action: Field;
-};
-
-type HttpsInputContext = BaseInputContext & {
-  action: string;
-};
+type PresentationRequestType = 'no-context' | 'zk-app' | 'https';
 
-type BaseWalletContext = {
+type WalletDerivedContext = {
+  vkHash: Field;
+  claims: Field;
   clientNonce: Field;
 };
 
-type ZkAppWalletContext = BaseWalletContext & {
-  verifierIdentity: PublicKey;
-};
-
-type HttpsWalletContext = BaseWalletContext & {
-  verifierIdentity: string;
-};
-
-type PresentationRequestType = 'no-context' | 'zk-app' | 'https';
-
 type PresentationRequest<
+  RequestType extends PresentationRequestType = PresentationRequestType,
   Output = any,
   Inputs extends Record<string, Input> = Record<string, Input>,
   InputContext = any,
   WalletContext = any
 > = {
-  type: PresentationRequestType;
+  type: RequestType;
   spec: Spec<Output, Inputs>;
   claims: Claims<Inputs>;
-  inputContext?: InputContext;
-  deriveContext(walletContext?: WalletContext): Field;
+  inputContext: InputContext;
+  program?: unknown;
+
+  deriveContext(
+    /**
+     * Context that is passed in from the input request / server-side
+     */
+    inputContext: InputContext,
+    /**
+     * Application-specific context that is passed in from the wallet / client-side
+     */
+    walletContext: WalletContext,
+    /**
+     * Context automatically (re-)derived on the client
+     */
+    derivedContext: WalletDerivedContext
+  ): Field;
 };
 
 const PresentationRequest = {
-  noContext<Output, Inputs extends Record<string, Input>>(
+  https<Output, Inputs extends Record<string, Input>>(
     spec: Spec<Output, Inputs>,
-    claims: Claims<Inputs>
+    claims: Claims<Inputs>,
+    context: { action: string }
   ) {
-    return {
-      type: 'no-context',
+    // generate random nonce on "the server"
+    let serverNonce = Field.random();
+
+    return HttpsRequest({
       spec,
       claims,
-      deriveContext: () => Field(0),
-    } satisfies PresentationRequest;
+      program: createProgram(spec),
+      inputContext: { type: 'https', ...context, serverNonce },
+    });
   },
 
   zkApp<Output, Inputs extends Record<string, Input>>(
     spec: Spec<Output, Inputs>,
     claims: Claims<Inputs>,
-    inputContext: ZkAppInputContext
-  ): PresentationRequest<Output, Inputs> {
-    const { presentationCircuitVKHash, action, serverNonce } = inputContext;
-    const claimsType = NestedProvable.fromValue(claims);
-    const claimsFields = Struct(claimsType).toFields(claims);
-    const claimsHash = Poseidon.hash(claimsFields);
```

Comment:
claims are now re-hashed on the client, no need to store the hash in input context